ledger/stateproofverificationtracker.go

```diff
@@ -0,0 +1,340 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/algorand/go-algorand/ledger/store"
+
+	"github.com/algorand/go-deadlock"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var (
+	errStateProofVerificationContextNotFound = errors.New("requested state proof verification context not found")
+)
+
+type verificationDeleteContext struct {
+	confirmedRound      basics.Round
+	stateProofNextRound basics.Round
+}
+
+type verificationCommitContext struct {
+	confirmedRound      basics.Round
+	verificationContext ledgercore.StateProofVerificationContext
+}
+
+// stateProofVerificationTracker is in charge of tracking context required to verify state proofs until such a time
+// as the context is no longer needed.
+type stateProofVerificationTracker struct {
+	// dbQueries are the pre-generated queries used to query the database, if needed,
+	// to lookup state proof verification context.
+	dbQueries *store.StateProofVerificationDbQueries
+
+	// trackedCommitContext represents the part of the tracked verification context currently in memory. Each element in this
+	// array contains both the context required to verify a single state proof and context to decide whether it's possible to
+	// commit the verification context to the database.
+	trackedCommitContext []verificationCommitContext
+
+	// trackedDeleteContext represents the context required to delete committed state proof verification context from the
+	// database.
+	trackedDeleteContext []verificationDeleteContext
+
+	// mu protects trackedCommitContext and trackedDeleteContext.
+	mu deadlock.RWMutex
+
+	// log copied from ledger
+	log logging.Logger
+
+	// lastLookedUpVerificationContext should store the last verification context that was looked up.
+	lastLookedUpVerificationContext ledgercore.StateProofVerificationContext
+}
+
+func (spt *stateProofVerificationTracker) loadFromDisk(l ledgerForTracker, _ basics.Round) error {
+	preparedDbQueries, err := store.StateProofVerificationInitDbQueries(l.trackerDB().Rdb.Handle)
+	if err != nil {
+		return err
+	}
+
+	spt.dbQueries = preparedDbQueries
+
+	spt.log = l.trackerLog()
+
+	spt.mu.Lock()
+	defer spt.mu.Unlock()
+
+	spt.lastLookedUpVerificationContext = ledgercore.StateProofVerificationContext{}
+
+	const initialContextArraySize = 10
+	spt.trackedCommitContext = make([]verificationCommitContext, 0, initialContextArraySize)
+	spt.trackedDeleteContext = make([]verificationDeleteContext, 0, initialContextArraySize)
+
+	return nil
+}
+
+func (spt *stateProofVerificationTracker) newBlock(blk bookkeeping.Block, delta ledgercore.StateDelta) {
```

Comment:
general node for this file: all names are tooooooooooo long it is really hard to read